pkg/registry/core/node/strategy.go

```diff
@@ -163,6 +181,22 @@ type nodeStatusStrategy struct {
 
 var StatusStrategy = nodeStatusStrategy{Strategy}
 
+// ResetFieldsFor returns a set of fields for the provided version that get reset before persisting the object.
+// If no fieldset is defined for a version, nil is returned.
+func (nodeStatusStrategy) ResetFieldsFor(version string) *fieldpath.Set {
+	set, ok := resetFieldsByVersionForStatus[version]
+	if !ok {
+		return nil
+	}
+	return set
+}
+
+var resetFieldsByVersionForStatus = map[string]*fieldpath.Set{
+	"v1": fieldpath.NewSet(
+		fieldpath.MakePathOrDie("spec"),
```

Comment:
you're missing `Status.Config` along with the conditions. I'm curious how we can test these or make sure we don't have any discrepancy (someone updating one list but not the other).

Reply:
@liggitt, do you know how we would be able to test that?

Reply:
hmm... nothing simple and comprehensive springs to mind... a generic test would have to do something like fuzz fields, then run it through PrepareForUpdate/PrepareForCreate, then diff to compute dropped/reset field paths? Create and update can behave differently, and update can depend on the previous value, which makes it difficult to use fuzzing.

Reply:
OK, one of the problem we're trying to fix is that we're only dealing with versioned objects in SMD, and the status-wiping in `PrepareForUpdate` is done on internal versions. As we convert objects in SMD to various versions for processing, I think we must go through internal version at some point, can we use that time to do the wiping, with a method that would be common to `PrepareForUdpate` and for us?

Reply:
You could use this list to do the wiping, remove the existing clearing code, and rely on the existing tests?

> you're missing Status.Config along with the conditions

I don't understand this, isn't this a list of fields to wipe when status is being updated?

Reply:
The wiping is done in `PrepareForUpdate` on an internal version, but this list is per-version?

> you're missing Status.Config along with the conditions

I meant that the list is incorrect, it's missing fields to wipe. And as long as we dont' have a good way to test that this list matches the actual wiping, it's probably going to be incorrect.

Reply:
But status.config and conditions should not be wiped?

Reply:
> The wiping is done in PrepareForUpdate on an internal version, but this list is per-version?

But perhaps it could be done prior to converting to the internal version?

Reply:
Where is the wiping currently tested?

Reply:
https://github.com/kubernetes/kubernetes/blob/d0e4eeafe1ba709385f096ad968f4c9bf9de4d91/pkg/registry/core/node/strategy.go#L205-L207

Shouldn't we also wipe `Status.Config` in that case?

Reply:
> > The wiping is done in PrepareForUpdate on an internal version, but this list is per-version?
>
> But perhaps it could be done prior to converting to the internal version?

@liggitt Would it be safe/easy to run `PrepareForUpdate` on a versioned object rather than internal?

Reply:
That's implementing a feature gate, it's not actually connected to status wiping.

Reply:
> Would it be safe/easy to run PrepareForUpdate on a versioned object rather than internal?

You may have to move it up or down the stack. In theory it is possible but in practice it may be very awkward.

Reply:
> That's implementing a feature gate, it's not actually connected to status wiping.

But still, would we want someone to own the alpha field that ends-up being "wiped"?

Reply:
It's very unclear that that's even a good way of doing it. I think it's acceptable for SSA to not be aware of fields cleared due to features being turned off, at least for the first pass. Note that that doesn't always clear the field, you can't put it in a list like this.

Reply:
> Note that that doesn't always clear the field, you can't put it in a list like this.

Absolutely agree. If we had this as a function rather than a list per version, then this would be much easier to do.

Reply:
> Would it be safe/easy to run PrepareForUpdate on a versioned object rather than internal?

It would only be possible by significantly reworking the handling stack, as Daniel said. It would not be safe or easy

Reply:
As far as I remember, the reason we're doing the per-version list rather than a split of `PrepareForUpdate` into `PrepareForUdpate` and `ResetFields` in the strategy is because it's using internal versions. Since we can (and do) convert objects into their internal versions for apply, why don't we just do that?

Reply:
We made the per-version list because we needed to pass something into the SSA code.

Reply:
This is getting long but ... since we have to do the wiping before going into anything SSA, we could still call that function before going to SSA?

Reply:
OK, Discussed with Daniel offline about that. If someone forgets to update the list, it's still better than the current situation, so we should probably move forward with that.